interop/src/main.rs

```diff
@@ -1,44 +1,166 @@
 use std::{
     net::{SocketAddr, ToSocketAddrs},
     sync::{Arc, Mutex},
+    time::Duration,
 };
 
-use anyhow::{anyhow, Error, Result};
-use futures::TryFutureExt;
+use anyhow::{anyhow, Result};
+use futures::{future, TryFutureExt};
+use lazy_static::lazy_static;
 use structopt::StructOpt;
-use tokio::{io::AsyncReadExt, runtime::Builder};
-use tracing::{info, warn};
+use tokio::io::AsyncReadExt;
+use tracing::{error, info, warn};
 
 #[derive(StructOpt, Debug)]
 #[structopt(name = "interop")]
 struct Opt {
-    host: String,
+    host: Option<String>,
     #[structopt(default_value = "4433")]
     port: u16,
     #[structopt(default_value = "4434")]
     retry_port: u16,
-
+    #[structopt(long)]
+    h3: bool,
+    #[structopt(long)]
+    hq: bool,
     /// Enable key logging
     #[structopt(long = "keylog")]
     keylog: bool,
 }
 
-fn main() {
-    let opt = Opt::from_args();
-    let code = {
-        tracing::subscriber::set_global_default(
-            tracing_subscriber::FmtSubscriber::builder()
-                .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
-                .finish(),
-        )
-        .unwrap();
-        if let Err(e) = run(opt) {
-            eprintln!("ERROR: {}", e);
-            1
-        } else {
-            0
+#[derive(Clone)]
+enum Alpn {
+    Hq,
+    H3,
+    HqH3,
+}
+
+impl From<&Alpn> for Vec<Vec<u8>> {
+    fn from(alpn: &Alpn) -> Vec<Vec<u8>> {
+        match alpn {
+            Alpn::H3 => vec![quinn_h3::ALPN.into()],
+            Alpn::Hq => vec![b"hq-24"[..].into()],
+            Alpn::HqH3 => vec![b"hq-24"[..].into(), quinn_h3::ALPN.into()],
+        }
+    }
+}
+
+#[derive(Clone)]
+struct Peer {
+    name: String,
+    host: String,
+    port: u16,
+    retry_port: u16,
+    alpn: Alpn,
+}
+
+impl Peer {
+    fn new<T: Into<String>>(host: T) -> Self {
+        let host_str = host.into();
+        Self {
+            name: host_str.clone(),
+            host: host_str,
+            port: 4433,
+            retry_port: 4434,
+            alpn: Alpn::HqH3,
         }
+    }
+
+    fn name<T: Into<String>>(mut self, name: T) -> Self {
+        self.name = name.into();
+        self
+    }
+
+    fn port(mut self, port: u16) -> Self {
+        self.port = port;
+        self
+    }
+
+    fn retry_port(mut self, port: u16) -> Self {
+        self.retry_port = port;
+        self
+    }
+
+    fn h3(mut self) -> Self {
+        self.alpn = Alpn::H3;
+        self
+    }
+
+    fn hq(mut self) -> Self {
+        self.alpn = Alpn::H3;
+        self
+    }
+}
+
+lazy_static! {
+    static ref PEERS: Vec<Peer> = vec![
```

Comment:
It's not terribly important, but if you modified `Peer` to use `const fn` and store `&str` rather than `String` then this could be `const PEERS: &[Peer<'static>]` instead.

Reply:
I tried a little in this direction, but then thought it wasn't worth to spend time on this. Isn't `const fn` still a nightly thing? Would it fit with constructing a Peer from cli args?

Reply:
`const fn` is stable, it just can't do much yet. Constructing structs is one of the things it can do. For example, the following [works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e38e803be2d18674cee2857efd924139):
```rust
struct Peer<'a> {
    host: &'a str
}

impl<'a> Peer<'a> {
    const fn new(host: &'a str) -> Self {
        Self { host }
    }
}

const PEERS: &[Peer<'static>] = &[
    Peer::new("foo")
];
```

For CLI-specified peers, you could place the arguments on the stack and end up with a non-static lifetime. If that's not viable, you could instead use `Cow<'static, str>` to handle both cases while still having a 'static struct.

Reply:
Ok, so I tried to write it like that, but I got problems chaining `const fn` functions.
I don't think it worth it to spend time on this for now...

That's been fun to try nevertheless, thanks.